chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2222 @@
+    if radio.radio_ident == radio.ident_Mode:
+        LOG.info("Successful match.")
+    else:
+        msg = (
+            "Please check whether the walkie-talkie mode is consistent with "
+            "the selected model mode.")
+        raise InvalidValueError(msg)
+
+    # Main block
+    LOG.info("Downloading...")
+
+    for i in range(0, radio._memsize, 0x20):
+        block = _read_block(radio, i, 0x20)
+        data += block
+        _do_status(radio, i)
+    _do_status(radio, radio._memsize)
+    LOG.info("done.")
+
+    if append_model:
+        data += radio.MODEL.ljust(8)
+
+    # LOG.debug("done.")
+    return memmap.MemoryMapBytes(data)
+
+
+def _exit_write_block(radio):
+    serial = radio.pipe
+    try:
+        serial.write(b"E")
+
+    except:
+        raise errors.RadioError("Radio refused to exit programming mode")
+
+
+def _write_block(radio, addr, data):
+    serial = radio.pipe
+    cmd = struct.pack(">cHb", b'W', addr, 0x20)
+    data = radio.get_mmap()[addr + 8: addr + 40]
+    cmd_str = hex(sum(data) & 0xff)
+    data_bused = chr(int(cmd_str, 16))
+
+    # This is because when our cmd_str is converted to decimal,
+    # sometimes num will be greater than 128, but chr(num).encode('utf-8') gets
+    # two-byte ASCII characters, each in ASCII encoding The character
+    # encoding is one byte, and only 0-127, and the encoding of the 128-255
+    # part belongs to Extended ASCII, and the characters in the extended
+    # ascii part greater than 128 are encoded by two bytes, and two
+    # characters will be obtained Bytes, what is needed is an
+    # encoding that supports Extended ASCII characters and uses 1 Byte
+    # fixed-size encoding, such as ISO 8859-1, also known as latin1:
+    # >>> chr(165)
+    # ¥
+    # >>> chr(165).encode('latin1')
+
+    data += data_bused.encode('latin1')
+    used_data = cmd + data
+    serial.write(used_data)
+
+    ack = radio.pipe.read(1)
+    if ack != b"\x06":
+        raise errors.RadioError("Radio refused to accept block 0x%04x" % addr)
+
+
+def _do_upload(radio):
+    ident = _ident_radio(radio)
+    radio_version = _get_radio_firmware_version(radio)
+    LOG.info("Radio Version is %s" % repr(radio_version))
+
+    # Main block
+    LOG.debug("Uploading...")
+
+    for start_addr, end_addr in radio._ranges_main:
+        for addr in range(start_addr, end_addr, 0x20):
+            _write_block(radio, addr, 0x20)
+            _do_status(radio, addr)
+    _exit_write_block(radio)
+    LOG.debug("Upload all done.")
+
+
+TX_POWER = [chirp_common.PowerLevel("Low",  watts=1.00),
+            chirp_common.PowerLevel("Mid",  watts=4.00),
+            chirp_common.PowerLevel("High", watts=8.00)]
+
+TDH8_CHARSET = chirp_common.CHARSET_UPPER_NUMERIC + \
+    "!@#$%^&*()+-=[]:\";'<>?,./"
+
+
+@directory.register
+class TDH8(chirp_common.CloneModeRadio):
+    """TIDRADIO TD-H8"""
+    VENDOR = "TIDRADIO"
+    MODEL = "TD-H8"
+    ident_Mode = "P31183"
```

Comment:
The "M" should not be capitalized, and the convention for model-specific stuff like this is to prefix it with an underscore, as you do elsewhere.